### PR TITLE
feat(aws-dynamodb-global): support for global dynamodb tables

### PR DESCRIPTION
**Pull request by KingOfPoptart**
----

### Pull Request Checklist

* [X] Testing
  - Unit test added (prefer not to modify an existing test, otherwise, it's probably a breaking change)
  - __CLI change?:__ coordinate update of integration tests with team
  - __cdk-init template change?:__ coordinated update of integration tests with team
* [x] Docs
  - __jsdocs__: All public APIs documented
  - __README__: README and/or documentation topic updated
* [X] Title and Description
  - __Change type__: title prefixed with **fix**, **feat** will appear in changelog
  - __Title__: use lower-case and doesn't end with a period
  - __Breaking?__: last paragraph: "BREAKING CHANGE: <describe what changed + link for details>"
  - __Issues__: Indicate issues fixed via: "**Fixes #xxx**" or "**Closes #xxx**"
* [ ] Sensitive Modules (requires 2 PR approvers)
  - IAM Policy Document (in @aws-cdk/aws-iam)
  - EC2 Security Groups and ACLs (in @aws-cdk/aws-ec2)
  - Grant APIs (only if not based on official documentation with a reference)

----

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license.

### CODE REVIEW

File: packages/@aws-cdk/aws-dynamodb-global/README.md [hunk 1]
```diff
@@ -0,0 +1,31 @@
+## @aws-cdk/aws-dynamodb-global
+
+Here is a minimal deployable Global DynamoDB tables definition:
+
+```javascript
```
Comment (sam-goodwin) [inline packages/@aws-cdk/aws-dynamodb-global/README.md]: typescript

File: packages/@aws-cdk/aws-dynamodb-global/lib/lambda-global-dynamodb.ts [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+    /**
+     * The CloudFormation CustomResource that will manage the lambda
+     */
+    public customResource: cfn.CustomResource;
+
+    constructor(scope: cdk.Construct, id: string, props: DynamoDBGlobalStackProps) {
+        super(scope, id, props);
+        const codeLocation = path.resolve(__dirname, "lambda", "handler.js.lambda");
+        this.lambdaFunctionContent = fs.readFileSync(codeLocation, "utf8");
+        this.lambdaFunction = new lambda.SingletonFunction(this, id + "-SingletonLambda", {
```
Comment (sam-goodwin) [inline packages/@aws-cdk/aws-dynamodb-global/lib/lambda-global-dynamodb.ts]: Why append to the `id`, you could just use `SingletonLambda` as the `id`.

File: packages/@aws-cdk/aws-dynamodb-global/lib/lambda-global-dynamodb.ts [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+
+    /**
+     * The CloudFormation CustomResource that will manage the lambda
+     */
+    public customResource: cfn.CustomResource;
+
+    constructor(scope: cdk.Construct, id: string, props: DynamoDBGlobalStackProps) {
+        super(scope, id, props);
+        const codeLocation = path.resolve(__dirname, "lambda", "handler.js.lambda");
+        this.lambdaFunctionContent = fs.readFileSync(codeLocation, "utf8");
```
Comment (sam-goodwin) [inline packages/@aws-cdk/aws-dynamodb-global/lib/lambda-global-dynamodb.ts]: Use `lambda.Code.asset(codeLocation)` instead of reading yourself.

File: packages/@aws-cdk/aws-dynamodb-global/lib/lambda-global-dynamodb.ts [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+        });
+        this.lambdaFunction.addToRolePolicy(new iam.PolicyStatement()
+            .allow()
+            .addAllResources()
+            .addAction("iam:CreateServiceLinkedRole")
+            .addAction("application-autoscaling:DeleteScalingPolicy")
+            .addAction("application-autoscaling:DeregisterScalableTarget")
+            .addAction("dynamodb:CreateGlobalTable")
+            .addAction("dynamodb:DescribeLimits")
+            .addAction("dynamodb:UpdateGlobalTable"));
```
Comment (sam-goodwin) [inline packages/@aws-cdk/aws-dynamodb-global/lib/lambda-global-dynamodb.ts]: The `dynamodb:*` permissions should be offered as static methods, either in this package or `@aws-cdk/aws-dynamodb`.

File: packages/@aws-cdk/aws-dynamodb-global/lib/lambda-global-dynamodb.ts [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+        this.customResource = new cfn.CustomResource(this, id + "-CfnCustomResource", {
```
Comment (sam-goodwin) [inline packages/@aws-cdk/aws-dynamodb-global/lib/lambda-global-dynamodb.ts]: Again, why append to the `id`?

File: packages/@aws-cdk/aws-dynamodb-global/package.json [hunk 6]
```diff
@@ -0,0 +1,81 @@
+{
+  "name": "@aws-cdk/aws-dynamodb-global",
```
Comment (sam-goodwin) [inline packages/@aws-cdk/aws-dynamodb-global/package.json]: Why do we need a new package as opposed to `@aws-cdk/aws-dynamodb`?

File: packages/@aws-cdk/aws-dynamodb-global/README.md [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+import globaldynamodb = require('@aws-cdk/aws-dynamodb-global');
+const CONSTRUCT_NAME = 'aws-cdk-dynamodb-global';
+const TABLE_PARTITION_KEY: Attribute = { name: 'hashKey', type: AttributeType.String };
+const STACK_PROPS: DynamoDBGlobalStackProps = {
```
Comment (sam-goodwin) [inline packages/@aws-cdk/aws-dynamodb-global/README.md]: This is not idiomatic, pass the properties directly when creating the construct:

```typescript
new GlobalTable(stack, 'aws-cdk-dynamodb-global', {
    dynamoProps: {
        partitionKey: TABLE_PARTITION_KEY,
        tableName: TABLE_NAME
    },
    regions: [ 'us-east-1', 'us-east-2', 'us-west-2' ],
    // etc.
});
```

File: packages/@aws-cdk/aws-dynamodb-global/README.md [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+    },
+    regions: [ 'us-east-1', 'us-east-2', 'us-west-2' ]
+};
+
+const stack = new Stack();
+const globalTables = new GlobalTable(stack, CONSTRUCT_NAME, STACK_PROPS);
+```
+
+## Implementation Notes
+Global Dynamodb tables is an odd case currently.  The way this package works -
```
Comment (sam-goodwin) [inline packages/@aws-cdk/aws-dynamodb-global/README.md]: Is there documentation on the recommended practice from AWS?

File: packages/@aws-cdk/aws-dynamodb-global/lib/lambda-global-dynamodb.ts (hunk 2 above)
Comment (KingOfPoptart) [reply packages/@aws-cdk/aws-dynamodb-global/lib/lambda-global-dynamodb.ts]: I found it easier to track the deployed stacks in CFN by having a common prefix in the stack ID.  Otherwise, its difficult to keep track

File: packages/@aws-cdk/aws-dynamodb-global/package.json (hunk 6 above)
Comment (KingOfPoptart) [reply packages/@aws-cdk/aws-dynamodb-global/package.json]: Made more sense to me to have a new package since it has different requirements than the regular aws-dynamodb table - i.e. required `tableName` and `streamSpecification`.  As well as its a higher level object than the dynamodb package.

Once CFN officially supports global-dynamodb tables, it would probably be easier to deprecate this package than remove it from an integrated portion of the aws-dynamodb package

File: packages/@aws-cdk/aws-dynamodb-global/README.md (hunk 8 above)
Comment (KingOfPoptart) [reply packages/@aws-cdk/aws-dynamodb-global/README.md]: AWS has no recommended practice for this since its not supported in CFN currently.  This is the solution determined by myself and by chatting with CDK contributors

File: packages/@aws-cdk/aws-dynamodb-global/lib/lambda-global-dynamodb.ts (hunk 3 above)
Comment (KingOfPoptart) [reply packages/@aws-cdk/aws-dynamodb-global/lib/lambda-global-dynamodb.ts]: `lambda.Code.asset(codeLocation)` requires a zip file - Keeping a zip file in the repo would require the maintainer to either remember to update the zip file up to date, or adding a step during the build and testing to zip that file up which adds a dependency on being able to zip up an arbitrary file.  

Keeping it as a read-in file simplifies that, unless you have a solution to make that simpler?

File: packages/@aws-cdk/aws-dynamodb-global/lib/lambda-global-dynamodb.ts (hunk 5 above)
Comment (KingOfPoptart) [reply packages/@aws-cdk/aws-dynamodb-global/lib/lambda-global-dynamodb.ts]: See above

**Comment by KingOfPoptart** [comment]
Changes have been posted to the PR, and responses to questions posted.

File: packages/@aws-cdk/aws-dynamodb-global/README.md [hunk 9]
```diff
@@ -0,0 +1,30 @@
+## @aws-cdk/aws-dynamodb-global
+
```
Comment (eladb) [inline packages/@aws-cdk/aws-dynamodb-global/README.md]: A few words what global table are

File: packages/@aws-cdk/aws-dynamodb-global/README.md [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+Here is a minimal deployable Global DynamoDB tables definition:
+
+```typescript
+import globaldynamodb = require('@aws-cdk/aws-dynamodb-global');
+const CONSTRUCT_NAME = 'aws-cdk-dynamodb-global';
```
Comment (eladb) [inline packages/@aws-cdk/aws-dynamodb-global/README.md]: These consts are not really adding much in terms of readability and they are only used once, just inline them

File: packages/@aws-cdk/aws-dynamodb-global/README.md [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+const TABLE_PARTITION_KEY: Attribute = { name: 'hashKey', type: AttributeType.String };
+
+const stack = new Stack();
+const globalTables = new GlobalTable(stack, CONSTRUCT_NAME, {
+    dynamoProps: {
```
Comment (eladb) [inline packages/@aws-cdk/aws-dynamodb-global/README.md]: Would make more sense for the GlobalTableProps to inherit from TableProps so you don't need `dynamoProps` as a separate nested structure (we try to avoid nesting as much as possible)

File: packages/@aws-cdk/aws-dynamodb-global/README.md [hunk 12]
```diff
@@ -0,0 +1,30 @@
...
+        partitionKey: TABLE_PARTITION_KEY,
+        tableName: TABLE_NAME
+    },
+    regions: [ 'us-east-1', 'us-east-2', 'us-west-2' ]
+});
+```
+
+## Implementation Notes
+Global Dynamodb tables is an odd case currently.  The way this package works -
```
Comment (eladb) [inline packages/@aws-cdk/aws-dynamodb-global/README.md]: Make sure you capitalize service names properly and use the prefix "Amazon" or "AWS" (in this case it's `Amazon DynamoDB`).

File: packages/@aws-cdk/aws-dynamodb-global/README.md [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+
+* Creates a DynamoDB table in a separate stack in each `DynamoDBGlobalStackProps.region` specified
+* Deploys a CFN Custom Resource to your stack's specified region that calls a lambda that runs the aws cli which calls `createGlobalTable()`
+
+### Notes
+
+`DynamoDBGlobalStackProps.dynamoProps` is essentially a copy of `dynamodb.TableProps` - however `tableName` is *required*.
```
Comment (eladb) [inline packages/@aws-cdk/aws-dynamodb-global/README.md]: You can extract all the props besides `tableName` from `TableProps` into ` TableOptions` interface and then have `GlobaProps` and `TableProps` inherit from `Options` (that's kind of how we do this)

File: packages/@aws-cdk/aws-dynamodb-global/README.md [hunk 14]
```diff
@@ -0,0 +1,30 @@
...
+
+GlobalTable() will override any set `dynamoProps.streamSpecification` to be `NEW_AND_OLD_IMAGES` since this is a required attribute for global dynamodb tables to work.
```
Comment (eladb) [inline packages/@aws-cdk/aws-dynamodb-global/README.md]: Probably should error if users specified something else

File: packages/@aws-cdk/aws-dynamodb-global/README.md (hunk 11 above)
Comment (KingOfPoptart) [reply packages/@aws-cdk/aws-dynamodb-global/README.md]: I ran into issues with trying to extend TableProps - I needed to do some fancy TypeScript-fu to change tableName from optional to required and jsii didn't accept the type change.  

The cleanest way I thought to make it work is to have GlobalDynamoDBProps just be a copy/paste of TableProps with just that change.  I'm open for specific suggestions on how to clean that up, though

File: packages/@aws-cdk/aws-dynamodb-global/README.md (hunk 13 above)
Comment (eladb) [reply packages/@aws-cdk/aws-dynamodb-global/README.md]:
```ts
// in @aws-cdk/aws-dynamodb

export interface TableOptions {
  // all props besides `tableName`
}

export interface TableProps extends TableOptions {
  tableName?: string;
}

// in aws-dynamodb-global
export interface GlobalStackProps extends TableOptions {
  tableName: string;
}
```

File: packages/@aws-cdk/aws-dynamodb-global/README.md (hunk 9 above)
Comment (KingOfPoptart) [reply packages/@aws-cdk/aws-dynamodb-global/README.md]: Done

File: packages/@aws-cdk/aws-dynamodb-global/README.md (hunk 11 above)
Comment (KingOfPoptart) [reply packages/@aws-cdk/aws-dynamodb-global/README.md]: I implemented this with your suggested change from below

**Comment by KingOfPoptart** [comment]
More changes pushed.

**Comment by KingOfPoptart** [comment]
bump

File: packages/@aws-cdk/aws-dynamodb-global/lib/aws-dynamodb-global.ts [hunk 15]
```diff
@@ -0,0 +1,71 @@
...
+     * to share across Global DynamoDB tables
+     */
+    tableName: string;
+}
+
+/**
+ * Properties for the mutliple DynamoDB tables to mash together into a
+ * global table
+ */
+export interface GlobalTableProps extends cdk.StackProps {
```
Comment (eladb) [inline packages/@aws-cdk/aws-dynamodb-global/lib/aws-dynamodb-global.ts]: Sorry, I probably didn't explain myself too well (interfaces can extend multiple bases):

```ts
export interface GlobalTableProps extends cdk.StackProps, dynamodb.TableOptions {
  tableName: string;
  regions: string[];
}
```

Then, you don't need `GlobalDynamoDBProps` at all.

We like flat ;-)

File: packages/@aws-cdk/aws-dynamodb-global/lib/aws-dynamodb-global.ts [hunk 16]
```diff
@@ -0,0 +1,71 @@
...
+    dynamoProps: GlobalDynamoDBProps;
+
+    /**
+     * Array of environments to create DynamoDB tables in
+     * Accounts should be omitted, or at least all identical
+     */
+    regions: string[];
+}
+
+export class GlobalTable extends cdk.Construct {
```
Comment (eladb) [inline packages/@aws-cdk/aws-dynamodb-global/lib/aws-dynamodb-global.ts]: Short description of this type (the README intro is great)

File: packages/@aws-cdk/aws-dynamodb-global/lib/index.ts [hunk 17]
```diff
@@ -0,0 +1,3 @@
+export * from "./aws-dynamodb-global";
+export * from "./lambda-global-dynamodb";
```
Comment (eladb) [inline packages/@aws-cdk/aws-dynamodb-global/lib/index.ts]: Do we need to export this module? Seems like an internal implementation

File: packages/@aws-cdk/aws-dynamodb-global/package.json (hunk 6 above)
Comment (eladb) [reply packages/@aws-cdk/aws-dynamodb-global/package.json]: I think it's fine to have this as a separate module. Maybe we can add a comment in the dynamodb README that mentions it

File: packages/@aws-cdk/aws-dynamodb-global/test/integ.dynamodb.global.ts [hunk 18]
```diff
@@ -0,0 +1,19 @@
+import { Attribute, AttributeType } from '@aws-cdk/aws-dynamodb';
+import { App } from '@aws-cdk/cdk';
+import { GlobalTable } from '../lib';
+
+// CDK parameters
+const STACK_NAME = 'global-dynamodb-integ';
+
+// DynamoDB table parameters
+const TABLE = 'GlobalTable';
```
Comment (eladb) [inline packages/@aws-cdk/aws-dynamodb-global/test/integ.dynamodb.global.ts]: Why are these consts needed?

File: packages/@aws-cdk/aws-dynamodb-global/lib/aws-dynamodb-global.ts [hunk 19]
```diff
@@ -0,0 +1,71 @@
...
+
+    /**
+     * Creates dynamoDB tables across regions that will be able to be globbed together into a global table
+     */
+    public tables: MultiDynamoDBStack[];
```
Comment (eladb) [inline packages/@aws-cdk/aws-dynamodb-global/lib/aws-dynamodb-global.ts]: I think the API should only expose the `Table` objects and not the stack that wraps them. Furthermore, I don't even thing you need this `MultiDynamoDBStack` class (and of course you don't need it in the public API). You can just create those stacks in-place:

```ts
for (const reg of props.regions) {
  const regionalStack = new cdk.Stack(this, '...', { env: { region: reg } });
  const regionalTable = new dynamodb.Table(regionalStack, 'Table', props);
  this.tables.push(regionalTable);
}
```

This is how you should define the property:

```ts
public readonly tables = new Array<dynamodb.Table>();
```

P.S. this is not ideal because the array is mutable. The "right" thing to do would be to return a clone:

```ts
private readonly _regionalTables = new Array<dynamodb.Table>();

public get regionalTables() {
  return this._regionalTables.map(x => x);
}
```

File: packages/@aws-cdk/aws-dynamodb-global/lib/aws-dynamodb-global.ts [hunk 20]
```diff
@@ -0,0 +1,71 @@
...
+     * Array of environments to create DynamoDB tables in
+     * Accounts should be omitted, or at least all identical
+     */
+    regions: string[];
+}
+
+export class GlobalTable extends cdk.Construct {
+
+    /**
+     * Creates dynamoDB tables across regions that will be able to be globbed together into a global table
```
Comment (eladb) [inline packages/@aws-cdk/aws-dynamodb-global/lib/aws-dynamodb-global.ts]: call this `regionalTables`

File: packages/@aws-cdk/aws-dynamodb-global/lib/lambda-global-dynamodb.ts [hunk 21]
```diff
@@ -0,0 +1,62 @@
...
+            lambdaProvider: this.lambdaFunction,
+            properties: {
+                regions: props.regions,
+                resourceType: "Custom::MakeGlobalDynamoDB",
+                tableName: props.dynamoProps.tableName,
+            },
+        });
+    }
+}
```
Comment (eladb) [inline packages/@aws-cdk/aws-dynamodb-global/lib/lambda-global-dynamodb.ts]: Please don't check in the handler.zip file! Where is the code and how is it bundled into a zip file?

File: packages/@aws-cdk/aws-dynamodb-global/lib/aws-dynamodb-global.ts (hunk 19 above)
Comment (davidbarsky) [reply packages/@aws-cdk/aws-dynamodb-global/lib/aws-dynamodb-global.ts]: Chiming in—if mutability is a concern, would it make sense to use a library like https://github.com/immutable-js/immutable-js, or would that be too big of a dependency to expose as a public API in the CDK?

Comment (eladb) [reply packages/@aws-cdk/aws-dynamodb-global/lib/aws-dynamodb-global.ts]: We prefer to avoid taking such dependencies (and frankly JS/TS are sufficient). Soon Typescript will have `x: readonly string[]` and the world will be whole.

File: packages/@aws-cdk/aws-dynamodb-global/lib/lambda-global-dynamodb.ts (hunk 21 above)
Comment (KingOfPoptart) [reply packages/@aws-cdk/aws-dynamodb-global/lib/lambda-global-dynamodb.ts]: That's been addressed now

File: packages/@aws-cdk/aws-dynamodb-global/lib/aws-dynamodb-global.ts [hunk 22]
```diff
@@ -0,0 +1,64 @@
+import dynamodb = require("@aws-cdk/aws-dynamodb");
+import cdk = require("@aws-cdk/cdk");
+import { LambdaGlobalDynamoDBMaker } from "./lambda-global-dynamodb";
+
+/**
+ * Properties for the mutliple DynamoDB tables to mash together into a
+ * global table
+ */
+export interface RegionalTablesProps extends cdk.StackProps, dynamodb.TableOptions {
+    /**
```
Comment (eladb) [inline packages/@aws-cdk/aws-dynamodb-global/lib/aws-dynamodb-global.ts]: 2 spaces please

File: packages/@aws-cdk/aws-dynamodb-global/lib/aws-dynamodb-global.ts [hunk 23]
```diff
@@ -0,0 +1,64 @@
...
+     * Enforces a particular table name to share across Global DynamoDB tables
```
Comment (eladb) [inline packages/@aws-cdk/aws-dynamodb-global/lib/aws-dynamodb-global.ts]: Wording here feels off. "Enforces?" Ideally if we can copy & paste from official AWS docs that would be better. I would write something like "Name of the DynamoDB table to use across all regional tables. This is required for global tables".

File: packages/@aws-cdk/aws-dynamodb-global/lib/aws-dynamodb-global.ts [hunk 24]
```diff
@@ -0,0 +1,64 @@
...
+
+    /**
+     * Array of environments to create DynamoDB tables in
+     * The tables will all be created in the same account
+     */
+    regions: string[];
+}
+
+/**
+ * Global Tables builds upon DynamoDB’s global footprint to provide you with a fully managed, multi-region,
```
Comment (eladb) [inline packages/@aws-cdk/aws-dynamodb-global/lib/aws-dynamodb-global.ts]: This describes Global Tables but the class is `RegionalTables`?

File: packages/@aws-cdk/aws-dynamodb-global/lib/aws-dynamodb-global.ts [hunk 25]
```diff
@@ -0,0 +1,64 @@
+import dynamodb = require("@aws-cdk/aws-dynamodb");
+import cdk = require("@aws-cdk/cdk");
+import { LambdaGlobalDynamoDBMaker } from "./lambda-global-dynamodb";
+
+/**
+ * Properties for the mutliple DynamoDB tables to mash together into a
+ * global table
+ */
+export interface RegionalTablesProps extends cdk.StackProps, dynamodb.TableOptions {
```
Comment (eladb) [inline packages/@aws-cdk/aws-dynamodb-global/lib/aws-dynamodb-global.ts]: Why is this called "RegionalTables" instead of "GlobalTable"?

File: packages/@aws-cdk/aws-dynamodb-global/lib/aws-dynamodb-global.ts [hunk 26]
```diff
@@ -0,0 +1,64 @@
...
+
+    /**
+     * Creates dynamoDB tables across regions that will be able to be globbed together into a global table
+     */
+    private readonly _regionalTables = new Array<dynamodb.Table>();
+
+    constructor(scope: cdk.Construct, id: string, props: RegionalTablesProps) {
+        super(scope, id);
+        this._regionalTables = [];
+        if ( props.streamSpecification != null && props.streamSpecification !== dynamodb.StreamViewType.NewAndOldImages) {
```
Comment (eladb) [inline packages/@aws-cdk/aws-dynamodb-global/lib/aws-dynamodb-global.ts]: nit: Remove space after "("

File: packages/@aws-cdk/aws-dynamodb-global/lib/aws-dynamodb-global.ts [hunk 27]
```diff
@@ -0,0 +1,64 @@
...
+            throw(new TypeError("dynamoProps.streamSpecification MUST be set to dynamodb.StreamViewType.NewAndOldImages"));
+        }
+        // Need to set this streamSpecification here, otherwise global tables don't work
+        // And no way to set a default value in an interface
+        props.streamSpecification = dynamodb.StreamViewType.NewAndOldImages;
```
Comment (eladb) [inline packages/@aws-cdk/aws-dynamodb-global/lib/aws-dynamodb-global.ts]: If you rebase from master you'll see that all props are now readonly, so this will need to be rewritten like:

```ts
props = { ...props, streamSpecification: ... }
```

File: packages/@aws-cdk/aws-dynamodb-global/lib/aws-dynamodb-global.ts [hunk 28]
```diff
@@ -0,0 +1,64 @@
...
+        for (const reg of props.regions) {
```
Comment (eladb) [inline packages/@aws-cdk/aws-dynamodb-global/lib/aws-dynamodb-global.ts]: A comment explaining what this for-loop is doing

File: packages/@aws-cdk/aws-dynamodb-global/lib/aws-dynamodb-global.ts [hunk 29]
```diff
@@ -0,0 +1,64 @@
...
+            this._regionalTables.push(regionalTable);
+        }
+
+        this.lambdaGlobalDynamodbMaker = new LambdaGlobalDynamoDBMaker(scope, id + "-GlobalDynamodbCustomResource", props);
+        for (const table of this._regionalTables) {
+            this.lambdaGlobalDynamodbMaker.customResource.node.addDependency(table);
+        }
+    }
+
+    public get regionalTables() {
```
Comment (eladb) [inline packages/@aws-cdk/aws-dynamodb-global/lib/aws-dynamodb-global.ts]: jsdoc

File: packages/@aws-cdk/aws-dynamodb-global/lib/lambda/handler.js [hunk 30]
```diff
@@ -0,0 +1,105 @@
...
+          sendResponse(event, context, "FAILED");
+        } 
+        else {
+          console.log(data);           // successful response
+          sendResponse(event, context, "SUCCESS", { Message: "Resource creation successful!" });
+        }
+      });
+    } else if (event.RequestType === "Update") {
+      console.log("UDPATE!");
+      // Put your custom update logic here
```
Comment (eladb) [inline packages/@aws-cdk/aws-dynamodb-global/lib/lambda/handler.js]: remove comment

File: packages/@aws-cdk/aws-dynamodb-global/lib/lambda/handler.js [hunk 31]
```diff
@@ -0,0 +1,105 @@
...
+      sendResponse(event, context, "SUCCESS", { Message: "Resource update successful!" });
```
Comment (eladb) [inline packages/@aws-cdk/aws-dynamodb-global/lib/lambda/handler.js]: Does it makes sense that update is a no-op? What if I add a region for example?

File: packages/@aws-cdk/aws-dynamodb-global/lib/lambda/handler.js [hunk 32]
```diff
@@ -0,0 +1,105 @@
...
+    } else if (event.RequestType === "Delete") {
+      console.log("DELETE!");
+      // Put your custom delete logic here
+      sendResponse(event, context, "SUCCESS", { Message: "Resource deletion successful!" });
+    } else {
+      console.log("FAILED!");
+      sendResponse(event, context, "FAILED");
+    }
```
Comment (eladb) [inline packages/@aws-cdk/aws-dynamodb-global/lib/lambda/handler.js]: You must catch all exceptions and send a FAILED response. Otherwise, any exception will cause the CFN deployment to timeout after 4 hours.

Then, all failure cases should just "throw".

Ideally, if you can also have only a single point in your code that sends a SUCCESS response, it would be safer, especially if someone updates this code in the future and forgets to handle some case (alternatively, "throw" when you fall through).

File: packages/@aws-cdk/aws-dynamodb-global/lib/lambda/handler.js [hunk 33]
```diff
@@ -0,0 +1,105 @@
...
+  };
+
+function setupWatchdogTimer(event, context, callback) {
```
Comment (eladb) [inline packages/@aws-cdk/aws-dynamodb-global/lib/lambda/handler.js]: Why is this needed?

File: packages/@aws-cdk/aws-dynamodb-global/lib/lambda/handler.js [hunk 34]
```diff
@@ -0,0 +1,105 @@
...
+      context.done();
+    });
+
+    request.on("error", function(error) {
+      console.log("sendResponse Error:" + error);
+      // Tell AWS Lambda that the function execution is done
+      context.done();
+    });
+
+    // write data to request body
```
Comment (eladb) [inline packages/@aws-cdk/aws-dynamodb-global/lib/lambda/handler.js]: Please add a unit test for this handler

File: packages/@aws-cdk/aws-dynamodb-global/test/test.lambda.handler.js [hunk 35]
```diff
@@ -0,0 +1,3 @@
+'use strict';
+
+const handler = require('../lib/lambda/handler');
```
Comment (eladb) [inline packages/@aws-cdk/aws-dynamodb-global/test/test.lambda.handler.js]: Not enough.. You'll have to mock stuff..

Comment (KingOfPoptart) [reply packages/@aws-cdk/aws-dynamodb-global/test/test.lambda.handler.js]: lol I know, just checked in what I had gotten done so far this morning :)

File: packages/@aws-cdk/aws-dynamodb-global/lib/aws-dynamodb-global.ts (hunk 25 above)
Comment (KingOfPoptart) [reply packages/@aws-cdk/aws-dynamodb-global/lib/aws-dynamodb-global.ts]: You asked for the class to be renamed?
https://github.com/awslabs/aws-cdk/pull/2082#discussion_r269786560

Comment (eladb) [reply packages/@aws-cdk/aws-dynamodb-global/lib/aws-dynamodb-global.ts]: Not the class, the property name. It's: `GlobalTable.regionalTables`. Otherwise it doesn't make sense...

**Comment by eladb** [comment]
@KingOfPoptart could you please rebase this and we'll merge that in

**Comment by KingOfPoptart** [comment]
This PR has been replaced by https://github.com/awslabs/aws-cdk/pull/2251
